### PR TITLE
Added systemd manager

### PR DESCRIPTION
**Pull request by Zyqsempai**
Added `NewSystemd` and `LoadSystemd`.
@AkihiroSuda @crosbymichael I am stuck a little bit here, please guide me in right direction;)
Signed-off-by: bpopovschi <zyqsempai@mail.ru>

### CODE REVIEW

File: v2/systemd.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	properties := []systemdDbus.Property{
+		systemdDbus.PropDescription(fmt.Sprintf("cgroup %s", name)),
+		systemdDbus.PropWants(slice),
+		newProperty("DefaultDependencies", false),
+		newProperty("MemoryAccounting", true),
+		newProperty("CPUAccounting", true),
+		newProperty("BlockIOAccounting", true),
+		newProperty("MemoryLimit", uint64(*resources.Memory.Limit)),
+		newProperty("CPUShares", uint64(*resources.CPU.Shares)),
+		newProperty("MemoryLimit", uint64(*resources.Memory.Limit)),
```
Comment (AkihiroSuda) [inline v2/systemd.go]: These are deprecated https://github.com/opencontainers/runc/issues/2164

File: v2/systemd.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+		return nil, err
+	}
+	path := filepath.Join(mountpoint, group)
+	return &Systemd{
+		unifiedMountpoint: mountpoint,
+		path:              path,
+	}, nil
+}
+
+type Systemd struct {
```
Comment (AkihiroSuda) [inline v2/systemd.go]: stats?

File: v2/systemd.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+const (
+	defaultSlice = "system.slice"
+)
+
+var (
+	canDelegate bool
+	once        sync.Once
+)
+
+func NewSystemd(mountpoint string, group string, resources *specs.LinuxResources) (*Systemd, error) {
```
Comment (AkihiroSuda) [inline v2/systemd.go]: why mountpoint?

Comment (AkihiroSuda) [reply v2/systemd.go]: and what is the convention of `group`?

**Comment by Zyqsempai** [comment]
Now when I am looking at the code, I am thinking maybe it's make more sense to add `systemd` management to the main `manager`?

**Comment by Zyqsempai** [comment]
Reworked it a little bit @AkihiroSuda PTAL

**Comment by Zyqsempai** [comment]
@AkihiroSuda Let's finish this one, PTAL

**Comment by AkihiroSuda** [comment]
can we have cgctl?

**Comment by Zyqsempai** [comment]
sure

**Comment by Zyqsempai** [comment]
@AkihiroSuda  PTAL

File: v2/manager.go [hunk 4]
```diff
@@ -550,3 +561,144 @@ func setDevices(path string, devices []specs.LinuxDeviceCgroup) error {
...
+	}
+
+	if resources.Memory != nil && *resources.Memory.Limit != 0 {
+		properties = append(properties,
+			newSystemdProperty("MemoryMax", uint64(*resources.Memory.Limit)))
+	}
+
+	if resources.CPU != nil && *resources.CPU.Shares != 0 {
+		properties = append(properties,
+			newSystemdProperty("CPUWeight", *resources.CPU.Shares))
```
Comment (AkihiroSuda) [inline v2/manager.go]: needs conversion?
https://github.com/containerd/cgroups/blob/fada802a7909d430bd17126fd6e4bbf5716f2bcd/v2/utils.go#L174-L177

File: v2/utils.go [hunk 5]
```diff
@@ -302,3 +305,64 @@ func toRdmaEntry(strEntries []string) []*stats.RdmaEntry {
...
+func isUnitExists(err error) bool {
+	if err != nil {
+		if dbusError, ok := err.(dbus.Error); ok {
+			return strings.Contains(dbusError.Name, "org.freedesktop.systemd1.UnitExists")
+		}
+	}
+	return false
+}
+
+func createCgroupsv2Path(path string) (Err error) {
```
Comment (AkihiroSuda) [inline v2/utils.go]: Why do you need this?

File: v2/utils.go [hunk 6]
```diff
@@ -302,3 +305,64 @@ func toRdmaEntry(strEntries []string) []*stats.RdmaEntry {
...
+				// If the directory was created, be sure it is not left around on errors.
+				defer func() {
+					if Err != nil {
+						os.Remove(current)
+					}
+				}()
+			}
+		}
+		if i < len(elements[3:])-1 {
+			if err := ioutil.WriteFile(filepath.Join(current, "cgroup.subtree_control"), resByte, defaultDirPerm); err != nil {
```
Comment (AkihiroSuda) [inline v2/utils.go]: this should use `ToggleControllers()`

**Comment by Zyqsempai** [comment]
@AkihiroSuda DONE

**Comment by Zyqsempai** [comment]
@AkihiroSuda ping

**Comment by AkihiroSuda** [comment]
How to test this with cgctl?

**Comment by Zyqsempai** [comment]
@AkihiroSuda @crosbymichael Soo, finally it look like something not so bad))
Please review it.
I added few tests for systemd in cpu and memory tests.
Also you can test it with cgctl by `make cgutil systemd test-sg.scope $pid` specify $pid you want to attach.

**Comment by AkihiroSuda** [comment]
Thanks, but Im on vacation without laptop. Will look into next year.

**Comment by Zyqsempai** [comment]
Have a nice holidays then;)

**Comment by Zyqsempai** [comment]
It's time to get back to work on this one;)

**Comment by Zyqsempai** [comment]
@crosbymichael @estesp PTAL, to merge this one finally;)

**Comment by crosbymichael** [comment]
LGTM

**Pull request merged**